immutable.go

```diff
@@ -2479,252 +2286,112 @@ func hashUint64(value uint64) uint32 {
-type uint16Comparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not an uint16.
-func (c *uint16Comparer) Compare(a, b interface{}) int {
-	if i, j := a.(uint16), b.(uint16); i < j {
-		return -1
-	} else if i > j {
-		return 1
-	}
-	return 0
-}
-
-// uint32Comparer compares two uint32 values. Implements Comparer.
-type uint32Comparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not an uint32.
-func (c *uint32Comparer) Compare(a, b interface{}) int {
-	if i, j := a.(uint32), b.(uint32); i < j {
-		return -1
-	} else if i > j {
-		return 1
-	}
-	return 0
-}
-
-// uint64Comparer compares two uint64 values. Implements Comparer.
-type uint64Comparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not an uint64.
-func (c *uint64Comparer) Compare(a, b interface{}) int {
-	if i, j := a.(uint64), b.(uint64); i < j {
-		return -1
-	} else if i > j {
-		return 1
-	}
-	return 0
-}
-
-// stringComparer compares two strings. Implements Comparer.
-type stringComparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not a string.
-func (c *stringComparer) Compare(a, b interface{}) int {
-	return strings.Compare(a.(string), b.(string))
-}
-
-// byteSliceComparer compares two byte slices. Implements Comparer.
-type byteSliceComparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not a byte slice.
-func (c *byteSliceComparer) Compare(a, b interface{}) int {
-	return bytes.Compare(a.([]byte), b.([]byte))
-}
-
 // reflectIntComparer compares two int values using reflection. Implements Comparer.
-type reflectIntComparer struct{}
+type reflectComparer[K constraints.Ordered] struct{}
 
 // Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
 // returns 0 if a is equal to b. Panic if a or b is not an int.
-func (c *reflectIntComparer) Compare(a, b interface{}) int {
-	if i, j := reflect.ValueOf(a).Int(), reflect.ValueOf(b).Int(); i < j {
-		return -1
-	} else if i > j {
-		return 1
-	}
-	return 0
-}
-
-// reflectUintComparer compares two uint values using reflection. Implements Comparer.
-type reflectUintComparer struct{}
-
-// Compare returns -1 if a is less than b, returns 1 if a is greater than b, and
-// returns 0 if a is equal to b. Panic if a or b is not an int.
-func (c *reflectUintComparer) Compare(a, b interface{}) int {
-	if i, j := reflect.ValueOf(a).Uint(), reflect.ValueOf(b).Uint(); i < j {
-		return -1
-	} else if i > j {
-		return 1
+func (c *reflectComparer[K]) Compare(a, b K) int {
```

Comment:
These comparers and hashers were a bit hard to unify. I ended up with 2 of each, but they could just as well have been one of each, or something like that...